File: test/integ/TypeAnalysisCallGraphGenerationTest.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <gtest/gtest.h>

#include "DexClass.h"
#include "GlobalTypeAnalysisPass.h"
#include "TypeAnalysisCallGraphGenerationPass.h"
#include "TypeAnalysisTestBase.h"

using namespace call_graph;
using namespace type_analyzer;
using namespace type_analyzer::global;

class TypeAnalysisCallGraphGenerationTest : public TypeAnalysisTestBase {};

TEST_F(TypeAnalysisCallGraphGenerationTest, Test) {
  auto scope = build_class_scope(stores);
  set_root_method(
      "Lcom/facebook/redextest/TypeAnalysisCallGraphGenerationTest;.main:()V");

  auto gta = new GlobalTypeAnalysisPass();
  auto cggen = new TypeAnalysisCallGraphGenerationPass();
  std::vector<Pass*> passes{gta, cggen};
  run_passes(passes);

  auto cg = cggen->get_result();
  ASSERT_TRUE(cg);

  auto meth_main =
      get_method("TypeAnalysisCallGraphGenerationTest;.main", "", "V");
  ASSERT_TRUE(meth_main);
  ASSERT_TRUE(cg->has_node(meth_main));

  // Check callees of invoke-virtual in meth_main.
  std::unordered_set<const DexMethod*> callees;
  for (auto const& s : GraphInterface::successors(*cg, cg->node(meth_main))) {
    const auto& target = GraphInterface::target(*cg, s);
    callees.insert(target->method());
  }
  ASSERT_TRUE(callees.count(get_method("Base;.getVal", "", "I")));
  ASSERT_TRUE(callees.count(get_method("SubOne;.getVal", "", "I")));
  ASSERT_TRUE(callees.count(get_method("SubTwo;.getVal", "", "I")));
}

File: test/integ/TypeAnalysisCallGraphGenerationTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

interface I {
  String getName();
}

class Base implements I {
  int getVal() { return 0; }
  @Override
  public String getName() { return "Base"; }
}

class SubOne extends Base implements I {
  @Override
  int getVal() {
    return 1;
  }
  @Override
  public String getName() {
    return "SubOne";
  }
}

class SubTwo extends Base implements I {
  @Override
  int getVal() {
    return 2;
  }
  @Override
  public String getName() {
    return "SubTwo";
  }
}

public class TypeAnalysisCallGraphGenerationTest {

  final Base mBase;
  final I mIntf;

  TypeAnalysisCallGraphGenerationTest() {
    mBase = new SubTwo();
    mIntf = new SubTwo();
  }

  int baseArg(Base b) {
    return b.getVal();
  }

  Base baseReturn(Base b) {
    if (b != null) {
      return b;
    }
    return null;
  }

  String intfArg(I i) { return i.getName(); }

  I intfReturn(I i) {
    if (i != null) {
      return i;
    }
    return null;
  }

  int baseField() { return mBase.getVal(); }

  String intfField() { return mIntf.getName(); }

  static void main() {
    Base b = new Base();
    b.getVal();

    TypeAnalysisCallGraphGenerationTest t =
        new TypeAnalysisCallGraphGenerationTest();
    SubOne s1 = new SubOne();
    b = t.baseReturn(s1);
    t.baseArg(b);

    I i = t.intfReturn(s1);
    t.intfArg(i);

    t.baseField();
    t.intfField();
  }
}
